apps/data-pipeline/library-traffic-scraper/src/lib.ts

```diff
@@ -0,0 +1,255 @@
+import type { database } from "@packages/db";
+import { libraryTraffic, libraryTrafficHistory } from "@packages/db/schema";
+import { conflictUpdateSetAllCols } from "@packages/db/utils";
+import { load } from "cheerio";
+import fetch from "cross-fetch";
+import { z } from "zod";
+
+const rawRespOKSchema = z.object({
+  message: z.literal("OK"),
+  data: z.object({
+    id: z.number(),
+    name: z.string(),
+    count: z.number(),
+    percentage: z.number(),
+    timestamp: z.string(),
+  }),
+});
+
+const rawRespSchema = z.union([
+  rawRespOKSchema,
+  z.object({ error: z.string() }),
+  z.object({ message: z.string() }),
+]);
+
+type RawRespOK = z.infer<typeof rawRespOKSchema>;
+
+export interface LocationMeta {
+  id: string;
+  libraryName: string;
+  locationLabel: string;
+  floorCode: string;
+}
+
+// Helper to sanitize and split raw JS array text content from HTML
+function parseArray(src: string): string[] {
+  return src
+    .split(/[,\n]/)
+    .map((s) => s.trim().replace(/^['"]|['"]$/g, ""))
+    .filter(Boolean);
+}
+
+// Build map of location metadata by ID from the UCI Libraries website
+async function collectLocationMeta(): Promise<Record<string, LocationMeta>> {
```

Comment:
Somewhat similarly to what has been added to #169, you should add code comments so that this code can be comfortably read without working through the exact behavior of each line of code, which for scraper code is especially tiring.